File: array/arch/1089-duplicate-zeros/duplicate-zeros.cc
/*
clang++ -std=c++20 duplicate-zeros.cc -o __a.out && ./__a.out
*/

#include <algorithm>
#include <vector>

using namespace std;

class Solution {
 public:
  void duplicateZeros(vector<int>& xs) {
    const int n = xs.size();
    int i = 0, j = 0;
    for (; j < n; ++i, ++j) {
      if (xs[i] == 0) {
        ++j;
      }
    }
    for (; i-- > 0;) {
      if (--j < n) {
        xs[j] = xs[i];
      }
      if (xs[i] == 0) {
        xs[--j] = 0;
      }
    }
  }
};

int main() {
  vector<int> xs{1, 0, 2, 3, 0, 4, 5, 0};
  Solution{}.duplicateZeros(xs);
  assert(vector<int>({1, 0, 0, 2, 3, 0, 0, 4}) == xs);
}

File: array/arch/1089-duplicate-zeros/duplicate-zeros.py
"""
python3 duplicate-zeros.py
"""

class Solution(object):
  def duplicateZeros(self, xs):
    """
    :type arr: List[int]
    :rtype: None Do not return anything, modify arr in-place instead.
    """

    n = len(xs)
    i = 0
    j = 0
    while j < n:
      if xs[i] == 0:
        j = j+1
      i = i+1
      j = j+1

    while i > 0:
      i = i-1
      j = j-1
      if j < n:
        xs[j] = xs[i]
      if xs[i] == 0:
        j = j-1
        xs[j] = 0


if __name__ == '__main__':
  xs = [1, 0, 2, 3, 0, 4, 5, 0]
  Solution().duplicateZeros(xs)
  assert [1, 0, 0, 2, 3, 0, 0, 4] == xs
